retrievalmarket/impl/provider.go

```diff
@@ -305,13 +313,48 @@ func (p *Provider) HandleQueryStream(stream rmnet.RetrievalQueryStream) {
 		if query.PieceCID != nil {
 			pieceCID = *query.PieceCID
 		}
-		pieceInfo, err := getPieceInfoFromCid(p.pieceStore, query.PayloadCID, pieceCID)
-
+		pieceInfo, isUnsealed, err := getPieceInfoFromCid(ctx, p.node, p.pieceStore, query.PayloadCID, pieceCID)
 		if err == nil && len(pieceInfo.Deals) > 0 {
 			answer.Status = retrievalmarket.QueryResponseAvailable
-			// TODO: get price, look for already unsealed ref to reduce work
 			answer.Size = uint64(pieceInfo.Deals[0].Length) // TODO: verify on intermediate
 			answer.PieceCIDFound = retrievalmarket.QueryItemAvailable
+
+			var storageDeals []abi.DealID
+			if pieceCID != cid.Undef {
+				// If the user wants to retrieve the payload from a specific piece, we only need to inspect storage deals
+				// made for that piece to quote a price for retrieval.
+				for _, d := range pieceInfo.Deals {
+					storageDeals = append(storageDeals, d.DealID)
+				}
+			} else {
+				// If the user does NOT want to retrieve the payload from a specific piece, we'll have to inspect all deals
+				// made for that piece to quote a price for retrieval.
+				storageDeals, err = getAllDealsContainingPayload(p.pieceStore, query.PayloadCID)
+				if err != nil {
+					log.Errorf("Retrieval query: getAllDealsContainingPayload: %s", err)
+					return
+				}
```

Comment:
It seems like in a few places in this function, if there is an error we log something but don't send QueryResponseError to the client. Can we restructure the code so that the client gets a response with an error message?
eg see the way the error response to GetMinerWorkerAddress() is handled above